### ISSUE TITLE
coroutinized noncopyable_function is not safe

### ISSUE THREAD

**Issue by xemul** [scylladb/seastar#1625]
Mixing chains with coroutines via noncopyable functions can be dangerous.

```c++
future<> call_later(noncopyable_function<future<>()> fn) {
    return seastar::yield().then([fn = std::move(fn)] () -> future<> {
        return fn();
    });
}
...
        return call_later([capture] () -> future<> {
            co_await seastar::yield();
            // capture is not safe to use here
        });
```

ASAN complain about head-/stack-use-after-free after the co-awaited yield in the callback. Wrapping anything with `coroutine::lambda` doesn't help, fn is released whenever then's fn returns. There are two fixes that can be applied here:

- copy capture on stack before yielding for the first time
- rewrite call_later's then continuation as `coroutine::lambda([] { co_await fn(); })`

**Comment by xemul** [comment]
Cc @avikivity , @piodul . Gents, please, comment

**Comment by avikivity** [comment]
It's not noncopyable_function's fault. future::then only guarantees stability during the immediate execution of func, not until its future completes. It would break even if call_later had the signature `future<> call_later(auto fn)` and was called with the lambda.

It's a mismatch between different styles. Continuations deal with address instability by bouncing parameters along continuations. Coroutines ensure stability instead. But if you pass a coroutine to a continuation, the expectations clash.

I don't think there's an easy way to fix it. A lambda coroutine signature (if translated to a free function) is `ret (const lambda&, args...)` indicating the value is passed by reference, so the lambda itself isn't captured. I asked for rvalue lambdas (`[capture] () &&`) but it didn't happen.

**Comment by dotnwat** [comment]
fwiw we found that even with coroutine::lambda it was too easy to make mistakes that were missed. we use clang-tidy in CI to reject all coroutine lambdas.

**Comment by xemul** [comment]
Is there a way to write concept/constraint that tells if `template <typename Func>`'s Func is a coroutine or not? Compiler should know it compile-time

**Comment by avikivity** [comment]
No, it's not exposed in the signature. Nor whether it's a lambda or not.
